src/keyring/KeyringController.test.ts

```diff
@@ -459,11 +616,64 @@ describe('KeyringController', () => {
     expect(signedTx).not.toBe('');
   });
 
+  it('should not sign transaction if from account is not provided', async () => {
+    let error = { message: '' };
+    try {
+      const account = initialState.keyrings[0].accounts[0];
+      const txParams = {
+        chainId: 3,
+        data: '0x1',
+        from: account,
+        gasLimit: '0x5108',
+        gasPrice: '0x5108',
+        to: '0x51253087e6f8358b5f10c0a94315d69db3357859',
+        value: '0x5208',
+      };
+      const unsignedEthTx = TransactionFactory.fromTxData(txParams, {
+        common: new Common(commonConfig),
+        freeze: false,
+      });
+      expect(unsignedEthTx.v).toBeUndefined();
+      await keyringController.signTransaction(unsignedEthTx, '');
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error.message).toContain(
+      'No keyring found for the requested account. Error info: The address passed in is invalid/empty; There are keyrings, but none match the address;',
+    );
+  });
+
+  /**
+   * TODO: There can be a better check in method signTransaction for valid transaction.
+   */
+  it('should not sign transaction if transaction is not valid', async () => {
+    let error = { message: '' };
+    try {
+      const account = initialState.keyrings[0].accounts[0];
+      const signedTx = await keyringController.signTransaction({}, account);
+      expect(signedTx.v).not.toBeUndefined();
+      expect(signedTx).not.toBe('');
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error.message).toContain('tx.sign is not a function');
+  });
+
   it('should submit password and decrypt', async () => {
     const state = await keyringController.submitPassword(password);
     expect(state).toStrictEqual(initialState);
   });
 
```

Comment:
`keyringController.submitPassword` returns current state even if wrong password is passed. This probably needs fix.

Reply:
Created issue https://github.com/MetaMask/controllers/issues/798 to look into it.